query_object.rb

```diff
@@ -1,45 +1,58 @@
-module YourSuperModule
+module ParamsMethods
   def self.included(host_class)
     host_class.extend(ClassMethods)
   end
 
   def initialize(params)
-    params.each do |k, v|
-      next eval("set_#{k}('#{v}')") if v.instance_of?(String)
-      eval("set_#{k}(#{v})")
-    end
+    @params = params
+  end
+
+  def query
+    params_for_query = available_params.map do |p|
+      "#{p}: #{value_for(p)}" if value_for(p)
+    end.compact
+    "Select with: #{params_for_query}"
+  end
+
+  def params
+    @params
+  end
+
+  def available_params
+    self.class.available_params
+  end
+
+  def default_value
+    self.class.default_value
+  end
+
+  def value_for(parameter)
+    params[parameter] || default_value[parameter]
   end
 
   module ClassMethods
-    def available_params(*params)
-      params.each do |parameter|
-        define_method(parameter) do
-          instance_variable_get("@#{parameter}") || (eval("default_#{parameter}") if respond_to?("default_#{parameter}"))
-        end
-
-        define_method("set_#{parameter}") do |value|
-          instance_variable_set("@#{parameter}", value)
-        end
-      end
+    def available_params(*available_params)
+      @available_params = available_params unless defined? @available_params
+      @available_params
     end
 
-    def default_value(*params)
-      define_method("default_#{params.first}") do
-        params.last
-      end
+    def default_value(*default_params)
+      @default_value = {} unless defined? @default_value
```

Comment:
How we can avoid to call this conditional every time?

Reply:
Fixed